lib/src/summation/sum.dart

```diff
@@ -0,0 +1,183 @@
+
+  /// Computes a [sum] across the provided [interfaces].
+  ///
+  /// The [width] can be either explicitly provided or inferred from other
+  /// values such as a [maxValue], [minValue], or [initialValue] that contain
+  /// width information (e.g. a [LogicValue]), or by making it large enough to
+  /// fit [maxValue], or by inspecting widths of [interfaces]. There must be
+  /// enough information provided to determine the [width].
+  ///
+  /// If no [maxValue] is provided, one will be inferred by the maximum that can
+  /// fit inside of the [width].
+  ///
+  /// It is expected that [maxValue] is at least [minValue], or else results may
+  /// be unpredictable.
+  Sum(
+    super.interfaces, {
+    dynamic initialValue = 0,
+    super.maxValue,
+    super.minValue,
+    super.width,
+    super.saturates,
+    super.name = 'sum',
+  }) : super(initialValue: initialValue) {
+    addOutput('sum', width: width);
+
+    var maxPosMagnitude = SummationBase.biggestVal(width);
+    var maxNegMagnitude = BigInt.zero;
+    for (final intf in interfaces) {
+      final maxMagnitude = intf.fixedAmount != null
+          ? intf.amount.value.toBigInt()
+          : SummationBase.biggestVal(intf.width);
+
+      if (intf.increments) {
+        maxPosMagnitude += maxMagnitude;
+      } else {
+        maxNegMagnitude += maxMagnitude;
+      }
+    }
+
+    // also consider that initialValue may be less than min or more than max
+    final maxInitialValueMagnitude = initialValue is Logic
+        ? SummationBase.biggestVal(initialValue.width)
+        : LogicValue.ofInferWidth(initialValue).toBigInt();
+    maxPosMagnitude += maxInitialValueMagnitude;
+    maxNegMagnitude += maxInitialValueMagnitude;
+
+    // calculate the largest number that we could have in intermediate
+    final internalWidth = max(
+        (maxPosMagnitude + maxNegMagnitude + BigInt.one).bitLength, width + 1);
+
+    final initialValueLogicExt = initialValueLogic.zeroExtend(internalWidth);
+    final minValueLogicExt = minValueLogic.zeroExtend(internalWidth);
+    final maxValueLogicExt = maxValueLogic.zeroExtend(internalWidth);
+
+    // lazy range so that it's not generated if not necessary
+    late final range = Logic(name: 'range', width: internalWidth)
+      ..gets(maxValueLogicExt - minValueLogicExt + 1);
+
+    final zeroPoint = Logic(name: 'zeroPoint', width: internalWidth)
+      ..gets(Const(maxNegMagnitude, width: internalWidth));
+
+    final upperSaturation = Logic(name: 'upperSaturation', width: internalWidth)
+      ..gets(maxValueLogicExt + zeroPoint);
+    final lowerSaturation = Logic(name: 'lowerSaturation', width: internalWidth)
+      ..gets(minValueLogicExt + zeroPoint);
+
+    final internalValue = Logic(name: 'internalValue', width: internalWidth);
+    sum <= (internalValue - zeroPoint).getRange(0, width);
+
+    final preAdjustmentValue =
+        Logic(name: 'preAdjustmentValue', width: internalWidth);
+
+    Combinational.ssa((s) => [
```

Comment:
I think this needs a comment as to what ssa is doing.  It's a little opaque.  I think you know because you made ssa actually work...

Reply:
sure, added some comments